src/test/fakes.rs

```diff
@@ -210,6 +210,15 @@ impl<'a> FakeRelease<'a> {
         self
     }
 
+    pub(crate) fn features(mut self, opt_features: Option<HashMap<String, Vec<String>>>) -> Self {
+        if let Some(features) = opt_features {
+            self.package.features = features;
+        } else {
+            self.package.features = HashMap::new();
+        }
```

Comment:
Wait, I'm confused - if the `features` field on package isn't optional, then how does this render pages for old docs? Does it confuse 'no features' with 'no data'? `[features]` with no entries is also valid, please add a test for that.

I ask because this definitely looks wrong - now there's two ways to say `HashMap::new()`, one by passing it in and one by passing in `None`.

Reply:
They are not optional, package will always have features that are passed from `cargo metadata`: `"features": []`. The option that I have enabled on the last version is that it will show it even for old docs. And feature flags page without features will state it. That seems to be desirable IMO.

Reply:
@almusil what about crates built before this feature was added? Won't it be `NULL` in the database?

Reply:
@jyn514 It will but the frontend part of this is optional. It is kinda confusing. Backend metadata are parsed as non-optional for the reason I have mentioned earlier. Frontend is optional so if the database part is NULL or simply empty it should result in the same thing AFAIK. Maybe I have wrong understanding of the templates and their evaluation of variables.

Reply:
Something weird is going on here, the tests don't match the live version. Notice here the features are optional: https://github.com/rust-lang/docs.rs/blob/2df55efcfde3541e83d0ebe2647a60cb6b266b5c/src/web/mod.rs#L525
but yet you're setting it here to a `HashMap` directly. I don't know why that's possible, but it's not right - we should be testing `None`, not just an empty HashMap.

> Maybe I have wrong understanding of the templates and their evaluation of variables.

It's not _wrong_ but you're not distinguishing between the two I think. `{% if metadata.features %}` is false *either* if the array is empty or if it doesn't exist. It would be nice to say 'no features' if it's empty and 'data not available' if it doesn't exist.

Reply:
That's the thing I don't know how to fake the NULL without changing the backend. The change in backend to have option would require some filtering if the features are just empty list from `cargo metadata`.

Reply:
Got it, that makes sense. In that case you can use a direct SQL query - create the package like normal, then `UPDATE releases SET features = NULL WHERE release_id = ?`.

Reply:
Ah right, I will add the test and update the frontend to match what you have described NULL vs empty. Thanks

Reply:
Should be done